wp-content/plugins/tapestry/controller/class.tapestry-controller.php

```diff
@@ -3,4 +3,110 @@
  * Add/update/retrieve Tapestry post and its child nodes
  * 
  */
-class TapestryController {}
\ No newline at end of file
+
+class TapestryController {
+    const POST_TYPES = [
+        'TAPESTRY' => 'tapestry',
+        'TAPESTRY_NODE' => 'tapestry_node'
+    ];
+    const ERRORS = [
+        'INVALID_POST_ID' => [
+            'MESSAGE' => 'PostID is invalid',
+            'STATUS' => ['status' => 404]
+        ]
+    ];
+
+    /**
+     * Update Tapestry nodes first then
+     * Update the existing Tapestry post if the postId is provided
+     * Otherwise, a new post will be created
+     * 
+     * @param type @post The post data
+     * @param type @postId The postId of the Tapestry
+     */
+    public function updateTapestry($tapestry, $postId = null) {
+        if (!isset($postId))
+            $postId = $this->insertPost($tapestry, 'tapestry');
+        $this->updateNodes($tapestry->nodes, $postId);
+
+        // TODO: Groups and Permisisons data could be added here later
+
+        if (!isset($tapestry->rootId))
+            $tapestry->rootId = $tapestry->nodes[0]->id;
+
+        $tapestry->nodes = $this->getNodeIds($tapestry->nodes);
+        update_post_meta($postId, 'tapestry', $tapestry);
+        return $tapestry;
+    }
+
+    /**
+     * Update Tapestry post's child nodes
+     * 
+     * @param type @nodes An array of child nodes to be updated
+     * @param type @postId The post ID to which the nodes belong
+     * @return WP_Error if postId is invalid
+     */
+    public function updateTapestryNodes($nodes, $postId = null) {
+        if (!isset($postId))
+            return $this->throwsError('INVALID_POST_ID');
+        $this->updateNodes($nodes, $postId);
+        return $nodes;
+    }
+
+    private function throwsError($code) {
```

Comment:
Please move all the helper functions to the bottom.

Reply:
I see. I updated.